Refactor repository into modules

isherman opened the pull request.

isherman commented on modules/core/README.md:
```diff
@@ -0,0 +1 @@
+- `FindGlog.cmake`. Copied from ceres? Vendor in source?
```
Use `FindGlog.cmake` from Ceres -- add header comments

isherman commented on cpp/farm_ng/capture_calibration_dataset.cpp:
```diff
@@ -8,25 +8,26 @@
-using farm_ng_proto::tractor::v1::ApriltagDetections;
-using farm_ng_proto::tractor::v1::BUCKET_CALIBRATION_DATASETS;
+using farm_ng_proto::perception_core::v1::ApriltagDetections;
+using farm_ng_proto::core::v1::BUCKET_CALIBRATION_DATASETS;
```
- remove `_proto`
- consider top-level `using` declaration to reduce verbosity

isherman commented on modules/perception_core/CMakeLists.txt:
```diff
@@ -0,0 +1,35 @@
+include_directories(${Protobuf_INCLUDE_DIRS})
+include_directories(${CMAKE_CURRENT_BINARY_DIR}/protos)
+
+# See https://stackoverflow.com/questions/42978414/
```
remove this; `perception_core` shouldn't be able to build independently

isherman replied:
eventually this would rely on `install` mechanism and `find_package(farm_ng_core)`

isherman commented on modules/perception_core/scripts/build-protos.sh:
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env bash
```
should this be replaced by CMakeLists? CMakeLists invokes Docker?

isherman commented on go/webrtc/cmd/proxy-server/main.go:
```diff
@@ -11,11 +11,11 @@ import (
 	"github.com/gorilla/mux"
 	"github.com/rs/cors"
 
+	corepb "github.com/farm-ng/core/genproto"
```
can we generate code (python, ts, go) into the build directory instead of into the source tree?

isherman replied:
done

isherman commented:
goals:
 - all code (in all languages) in modules
 - no need for modules to build independently
 - top-level build script that builds all non-cpp protobufs out-of-tree
 - centralize macros for cmakelists
 - centralize CXX flags

isherman commented on modules/core/cmake/FindGlog.cmake:
```diff
@@ -0,0 +1,158 @@
+# Copied from:
```
- top-level cmakelists.txt will add find scripts to the cmake_modules_path, set global CXX flags

ethanrublee commented on modules/calibration/protos/farm_ng/v1/calibrate_base_to_camera.proto:
```diff
@@ -1,18 +1,18 @@
 
-package farm_ng_proto.tractor.v1;
-option go_package = "github.com/farm_ng/genproto";
+package farm_ng.v1;
```
+1

isherman commented:
For discussion; how is `.env` used?

isherman commented:
[Now supports](https://github.com/farm-ng/tractor/pull/153/files#diff-b903e9728d231dc2aa9833ef7623b9f5524f7ea756b032a2af7162f1f66b3f3e) releasing module source to `tar.gz` files in a Github release

(for example, with dummy contents: https://github.com/isherman/tractor/releases/tag/v0.0.5-ian-temp)

isherman commented:
The following is required for the new `go get -u ...` protobuf generators to be installed:
```
export GOPATH=~/go
export PATH=$PATH:$GOPATH/bin
```

- [ ] Set up GOPATH to be in `env/`?

ethanrublee commented on modules/calibration/cpp/farm_ng/tracking_camera.cpp:
```diff
@@ -402,5 +400,6 @@ int Main(farm_ng::EventBus& bus) {
 }
 
 int main(int argc, char* argv[]) {
+  std::cerr << "And here." << std::endl;
```
doh, remove.

isherman replied:
done.

ethanrublee commented on modules/perception_core/cpp/farm_ng/perception_core/frame_grabber_intel.cpp:
```diff
@@ -0,0 +1,183 @@
+          new FrameGrabberIntel(event_bus, config));
+    };
+
+static int _t265 = FrameGrabber::AddFrameGrabberFactory(
```
@isherman here at runtime, during static initialization, the factory functions get registered with frame_grabber.

isherman replied:
i see

ethanrublee commented on setup.bash:
```diff
@@ -34,13 +34,14 @@ export PATH=$PATH:/usr/local/go/bin
-  fi
-done
+
+# # Bootstrap blobstore if necessary
```
This was causing errors, commented, need to move this somewhere else.

isherman commented:
- [ ] Support enabling/disabling proto codegen by language (to keep Dockerfiles simple)

isherman commented:
@ethanrublee since our last conversation I
- added cpp namespaces
- added per-language proto targets (to support C++-only docker builds) 
- renamed some things (e.g. TrackingCamera -> CameraPipeline)
- deleted some things (e.g. CaptureCalibrationDataset)

`docker-compose build && docker-compose up` is working for me. I think we're ready for a review / merge.

Only remaining issue i'm aware of is https://github.com/farm-ng/tractor/pull/153#issuecomment-730584463; looking at that now.
